terrain-manager.js

```diff
@@ -0,0 +1,180 @@
+		);
+
+		this.geometry = new THREE.BufferGeometry();
+
+		this.indexAttribute = new THREE.Uint32BufferAttribute();
+		this.indexAttribute.array = this.bufferFactory.indices;
+		this.indexAttribute.itemSize = 1;
+		this.indexAttribute.count = this.bufferFactory.indices.length;
+		this.indexAttribute.setUsage( THREE.DynamicDrawUsage );
+
+		this.positionAttribute = new THREE.Float32BufferAttribute();
+		this.positionAttribute.array = this.bufferFactory.positions;
+		this.positionAttribute.itemSize = 3;
+		this.positionAttribute.count = this.bufferFactory.positions.length / 3;
+		this.positionAttribute.setUsage( THREE.DynamicDrawUsage );
+
+		this.normalAttribute = new THREE.Float32BufferAttribute();
+		this.normalAttribute.array = this.bufferFactory.normals;
+		this.normalAttribute.itemSize = 3;
+		this.normalAttribute.count = this.bufferFactory.normals.length / 3;
+		this.normalAttribute.setUsage( THREE.DynamicDrawUsage );
+
+		this.geometry.setIndex(this.indexAttribute);
+		this.geometry.setAttribute('position', this.positionAttribute);
+		this.geometry.setAttribute('normal', this.normalAttribute);
+
+		this.geometry.clearGroups();
+
+		for (let i = 0; i < this.chunkCount * this.chunkCount; i++) {
+			this.geometry.addGroup(
+				this.bufferFactory.indexRanges[2 * i], this.bufferFactory.indexRanges[2 * i + 1], 0
+			);
+		}
+
+		this.mesh = new THREE.Mesh(
+			this.geometry, [new THREE.MeshLambertMaterial({ color: 0xff0000, wireframe: true })]
+		);
+	}
+
+	_calculateTargetChunks() {
+
+		let centerChunkGridX = Math.floor(this.center.x / this.chunkSize);
+		let centerChunkGridZ = Math.floor(this.center.z / this.chunkSize);
+
+		let targetChunks = [];
+
+		for (let i = centerChunkGridX - this.chunkRange; i < centerChunkGridX + this.chunkRange + 1; i++) {
+			for (let j = centerChunkGridZ - this.chunkRange; j < centerChunkGridZ + this.chunkRange + 1; j++) {
+				targetChunks.push(i + ':' + j);
+			}
+		}
+
+		return targetChunks;
+	}
+
+	updateCenter(pos) {
+
+		this.center = pos;
+
+		this.targetChunkIds = this._calculateTargetChunks();
+	}
+
+	updateChunk() {
+
+		const buf = this.bufferFactory;
+
+		let chunkIdToAdd = this.targetChunkIds.filter(
+			id => !this.currentChunks.map(v => v.chunkId).includes(id)
+		).at(0);
+
+		console.log(">>> chunk id to add: ", chunkIdToAdd);
+
+		let chunksToRemove = this.currentChunks.filter(chunk => !this.targetChunkIds.includes(chunk.chunkId));
+
+		console.log(">>> chunk id to remove: ", chunksToRemove);
+
+		chunksToRemove.forEach(chunk => {
+			this.geometryUtils.deallocateChunk(
+				chunk.slots[0], chunk.slots[1], this.chunkCount ** 2,
+				buf.chunkVertexRangeBuffer,
+				buf.vertexFreeRangeBuffer,
+				buf.chunkIndexRangeBuffer,
+				buf.indexFreeRangeBuffer
+			);
+		});
+
+		this.currentChunks = this.currentChunks.filter(chunk => this.targetChunkIds.includes(chunk.chunkId));
+
+		if (!!chunkIdToAdd) {
+			let gridId = chunkIdToAdd.split(':');
+
+			let slots = this.geometryUtils.allocateChunk(
```

Comment:
These WASM calls can be done in a worker.

Yes, we might have to pass back and forth transferables, or use `SharedArrayBuffer`. But it would still be a lot faster.

Reply:
Does `allocateChunk` allocate a chunk? Or generate it? Or both?

This seems to be called on update so I guess it would be both generating and allocating chunks.

So should it be called `generateChunk`?

Reply:
Yeah. It's doing both generating and allocating